From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: sf2000-linux <sf2000-linux@local>
Date: Wed, 13 May 2026 00:00:00 +0000
Subject: [PATCH] skip mips gp setup for flat crt1

The MIPS ELF crt1 startup assumes the linker script provides _gp for
non-PIC binaries.  elf2flt's FLAT link path does not provide that symbol,
and the SF2000 NOMMU userspace is built without small-data sections.  Do
not emit the _gp load for FLAT startup objects.
---
 libc/sysdeps/linux/mips/crt1.S | 4 +++-
 1 file changed, 3 insertions(+), 1 deletion(-)

diff --git a/libc/sysdeps/linux/mips/crt1.S b/libc/sysdeps/linux/mips/crt1.S
index b841db099..28c4af458 100644
--- a/libc/sysdeps/linux/mips/crt1.S
+++ b/libc/sysdeps/linux/mips/crt1.S
@@ -90,9 +90,11 @@ __start:
 #ifdef __PIC__
 	SETUP_GPX($0)
 	SETUP_GPX64($25,$0)
-#else
+#elif !defined(__UCLIBC_FORMAT_FLAT__)
 	PTR_LA $28, _gp             /* Setup GP correctly if we're non-PIC.  */
 	move $31, $0
+#else
+	move $31, $0
 #endif
 #ifdef L_rcrt1
 	PTR_LA $4, _DYNAMIC           /* Place _DYNAMIC into the GOT */
-- 
2.50.0
